synapse/static/client/login/js/login.js

```diff
@@ -46,23 +48,36 @@ var setFeedbackString = function(text) {
 };
 
 var show_login = function() {
-    $("#loading").hide();
-
     var this_page = window.location.origin + window.location.pathname;
     $("#sso_redirect_url").val(this_page);
 
-    if (matrixLogin.serverAcceptsPassword) {
-        $("#password_flow").show();
+    if (matrixLogin.serverAcceptsSso) {
+        if (try_token() || matrixLogin.serverAcceptsPassword) {
+            // Show the SSO form if there's a login token in the query. That's because,
+            // if there is a token, and this function is run, it means an error happened,
+            // and in this case it's nicer to show the form with an error rather than
+            // redirect immediately to the SSO portal.
+            // Also show the form if the server is accepting password login as well.
+            $("#sso_form").show();
+        } else {
+            // Submit the SSO form instead of displaying it. The reason behind this
+            // behaviour is that the user will likely arrive here after clicking on a
+            // button, in the client, with a label such as "Continue with SSO", so
+            // clicking on another button with the same semantics is a bit pointless.
+            $("#sso_form").submit();
+            return;
+        }
```

Comment:
I feel like this logic would make more sense if it it were reversed:

```js
    # If inhibit_redirect is false, and SSO is the only supported login method, we can
    # redirect straight to the SSO page
    if (!inhibit_redirect && !matrixLogin.serverAcceptsPassword) {
            $("#sso_form").submit();
            return;
    }

    # otherwise, show the SSO form the form
    $("#sso_form").show();
```